src/hashfunctions.c

```diff
@@ -0,0 +1,273 @@
+        return HASHKEY_BAG_NC(squash, 1, 0, max_point * 2);
+    }
+    else {
+        return HASHKEY_BAG_NC(perm, 1, 0, max_point * 4);
+    }
+}
+Obj DATA_HASH_FUNC_FOR_PERM(Obj self, Obj perm)
+{
+    /* check the argument                                                  */
+    if (TNUM_OBJ(perm) != T_PERM2 && TNUM_OBJ(perm) != T_PERM4) {
+        ErrorMayQuit("DATA_HASH_FUNC_FOR_PERM: <perm> must be a permutation "
+                     "(not a %s)",
+                     (Int)TNAM_OBJ(perm), 0L);
+    }
+
+    return HashValueToObjInt(DataHashFuncForPerm(perm));
+}
+
+Obj DATA_HASH_FUNC_FOR_PPERM(Obj self, Obj pperm)
+{
+    /* check the argument                                                  */
+    if (!IS_PPERM(pperm)) {
+        ErrorMayQuit("DATA_HASH_FUNC_FOR_PPERM: <pperm> must be a "
+                     "partial permutation (not a %s)",
+                     (Int)TNAM_OBJ(pperm), 0L);
+    }
+
+    return HashValueToObjInt(HashFuncForPPerm(pperm));
+}
+
+Obj DATA_HASH_FUNC_FOR_TRANS(Obj self, Obj trans)
+{
+    /* check the argument                                                  */
+    if (!IS_TRANS(trans)) {
+        ErrorMayQuit("DATA_HASH_FUNC_FOR_TRANS: <trans> must be a "
+                     "transformation (not a %s)",
+                     (Int)TNAM_OBJ(trans), 0L);
+    }
+
+    return HashValueToObjInt(HashFuncForTrans(trans));
+}
+
+Obj DATA_HASH_FUNC_FOR_STRING(Obj self, Obj string)
+{
+    /* check the argument                                                  */
+    if (!IS_STRING(string)) {
+        ErrorMayQuit("DATA_HASH_FUNC_FOR_STRING: <string> must be a "
+                     "string (not a %s)",
+                     (Int)TNAM_OBJ(string), 0L);
+    }
+
+    if(!IS_STRING_REP(string))
+    {
+        string = CopyToStringRep(string);
+    }
+
+
+    UInt len = GET_LEN_STRING(string);
+    UInt1* ptr = CHARS_STRING(string);
+
+    UInt hashval = HASHKEY_MEM_NC(ptr, 2782, len);
+    return HashValueToObjInt(hashval);
+}
+
+Int DataHashFuncForInt(Obj i)
+{
+    GAP_ASSERT(TNUM_OBJ(i) == T_INTPOS || TNUM_OBJ(i) == T_INTNEG);
+    if (TNUM_OBJ(i) == T_INTPOS)
+    {
+        return HASHKEY_WHOLE_BAG_NC(i, 293479);
+    }
+    else
+    {
+        return HASHKEY_WHOLE_BAG_NC(i, 193492);
+    }
+}
+
+Obj DATA_HASH_FUNC_FOR_INT(Obj self, Obj i)
+{
+    /* check the argument                                                  */
+    if (TNUM_OBJ(i) != T_INT && TNUM_OBJ(i) != T_INTPOS &&
+        TNUM_OBJ(i) != T_INTNEG) {
+        ErrorMayQuit(
+            "DATA_HASH_FUNC_FOR_INT: <i> must be an integer (not a %s)",
+            (Int)TNAM_OBJ(i), 0L);
+    }
+
+    if (IS_INTOBJ(i))
+        return i;
+    else
+        return HashValueToObjInt(DataHashFuncForInt(i));
+}
+
+Int BasicRecursiveHash(Obj obj);
+
+Int BasicRecursiveHashForList(Obj obj)
+{
+    GAP_ASSERT(IS_LIST(obj));
+    // This is just a randomly chosen number which fits in a 32-bit integer
+    UInt current_hash = 2195952830;
```

Comment:
So here you have a comment for the origin of the magic constant; but nowhere else -> having those elsewhere, too, would have been nice...